pkgs/applications/virtualization/xen/generic.nix

```diff
@@ -1,265 +1,376 @@
+        bison
+        bzip2
+        checkpolicy
+        dev86
+        figlet
+        flex
+        gettext
+        glib
+        acpica-tools
+        libaio
+        libiconv
+        libuuid
+        ncurses
+        openssl
+        perl
+        python311Packages.python
+        xz
+        yajl
+        zlib
+
+        # oxenstored
+        ocamlPackages.findlib
+        ocamlPackages.ocaml
+        systemd
+
+        # Python fixes
+        python311Packages.wrapPython
+
+        # Documentation
+        python311Packages.markdown
+        fig2dev
+        ghostscript
+        texinfo
+        pandoc
+
+        # Others
+      ]
+      ++ (lib.lists.concatMap (x: x.buildInputs or [ ]) (builtins.attrValues config.prefetchedSources))
+      ++ (config.buildInputs or [ ]);
+
+    prePatch =
+      # Xen's stubdoms, tools and firmwares need various sources that
+      # are usually fetched at build time using wget and git. We can't
+      # have that, so we pre-fetch them in the versioned Nix expressions,
+      # and produce fake wget and git executables for debugging purposes.
+      #
+      # The snippet below produces executables that simply print in stdout
+      # what they were supposed to fetch, and exit gracefully.
+      ''
+        mkdir fake-bin
+
+        cat > fake-bin/wget << EOF
+        #!${stdenv.shell} -e
+        echo ===== FAKE WGET: Not fetching \$*
+        [ -e \$3 ]
+        EOF
+
+        cat > fake-bin/git << EOF
+        #!${stdenv.shell}
+        echo ===== FAKE GIT: Not cloning \$*
+        [ -e \$3 ]
+        EOF
+
+        chmod +x fake-bin/*
+        export PATH=$PATH:$PWD/fake-bin
+      ''
+
+      # Remove in-tree QEMU sources, as we either pre-fetch them through
+      # the versioned Nix expressions if withInternalQEMU is true, or we
+      # don't build QEMU at all if withInternalQEMU is false.
+      + ''
+        rm -rf tools/qemu-xen tools/qemu-xen-traditional
+      ''
+
+      # The following expression moves the sources we fetched in the
+      # versioned Nix expressions to their correct locations inside
+      # the Xen source tree.
+      + ''
+        ${withPrefetchedSources (
+          name: pkg: ''
+            echo "Copying pre-fetched source: ${pkg.src} -> tools/${name}"
+            cp -r ${pkg.src} tools/${name}
+            chmod -R +w tools/${name}
+          ''
+        )}
+      '';
+
+    patches = config.patches or [ ];
+
+    postPatch =
+      # The following patch forces Xen to install xen.efi on $out/boot
+      # instead of $out/boot/efi/efi/nixos, as the latter directory
+      # would otherwise need to be created manually.
+      ''
+        substituteInPlace xen/Makefile \
+          --replace-fail "\$(D)\$(EFI_MOUNTPOINT)/efi/\$(EFI_VENDOR)/\$(T)-\$(XEN_FULLVERSION).efi" \
+                    "\$(D)\$(BOOT_DIR)/\$(T)-\$(XEN_FULLVERSION).efi"
+      ''
+
+      # The following patch fixes the call to /bin/mkdir on the
+      # launch_xenstore.sh helper script.
```

Comment:
Fix this upstream? Seems like a bug to me.

Reply:
Are you sure? It's a Nix-specific thing. Most sane environments will have /bin/mkdir, it's just that our mkdir is somewhere in the Nix Store instead.

Reply:
> Are you sure? It's a Nix-specific thing. Most sane environments will have /bin/mkdir, it's just that our mkdir is somewhere in the Nix Store instead. -- @SigmaSquadron (https://github.com/NixOS/nixpkgs/pull/324693#discussion_r1685814201)

Any script that is environment-dependent is in my mind flawed. 

`/bn/mkdir` is not a sane declaration, shell is designed to use PATH and be environment independent so that should be used instead.

Reply:
CC @andyhhp: Would Xen be interested in such a patch?

Reply:
@SigmaSquadron Ping me if you do, i would like to review it since it's a fragile shell script that will probably end up running on [NiXium](https://github.com/NiXium-org/NiXIum)